text/0000-undo-universal-impl-trait.md

```diff
@@ -0,0 +1,238 @@
+callee. The initial assumption that having a different mechanism is confusing is wrong because
+`impl Trait` carries its own semantics and because anonymization in return type is not a trivial
+feature people will use without actually wanting it.
+
+Also, people are used to generics. People coming from C++, C#, Java etc. actually know about
+generics. The links listed above prove how confusing it can get to show them a complete, different
+way to use generics (`impl Trait` in argument position is a use of hidden generics, that is,
+anonymized type variables with explicit trait bounds). For people who don’t come from such
+languages or for completely newcomers, we are assuming that generics are too hard to understand at
+first but people will eventually need to learn them, as `impl Trait` in argument position is
+strictly less powerful than type variables and trait bounds. They will have to read the book –
+perhaps several times. It then confusing to be introduced a concept that is less powerful than
+generics and yet used in a complete, separate way that has no other alternative (return location).
+
+Finally, advanced Rustaceans are also quite confused, because they’ve been using `impl Trait` in
+return location for a while now (mostly on nightly) and the initial RFC was well accepted among the
+community and was a success. Those developers obviously mapped `impl Trait` to *callee-picked*
+types, because at that time it was its sole semantics. When a function with an argument annotated
+with `impl Trait` appears to such a developer, they get confused because that would mean *an argument
+which type is chosen by the callee*, which doesn’t quite make sense.
+
+### Argument from ergonomics
+
+> Argument [here](https://github.com/rust-lang/rfcs/blob/master/text/1951-expand-impl-trait.md#argument-from-ergonomics)
+
+In this argument, [RFC 1951] states that having `impl Trait` in argument position eases how our
+mental model about a function works. We don’t have to read the type variables and rembember their
+trait bounds anymore to understand the contract of a function.
+
+This argument seems sound but again is wrong. Consider:
+
+```
+fn foo<T>(a: T, b: T, c: impl Debug) where T: Add<Output = T> + Copy
+```
+
+This function has three type variables to substitute to be completely monomorphized. However, it’s
+very hard to see it at first because one of the variables is hidden. Worse, the ergonomics argument
+doesn’t stand here because you can see you **have** to read the type variables in order to get what
+you can do with the variables.
+
+Also, consider this function that takes an argument and add it to itself:
+
+```
+fn add_self<T>(x: T) -> T where T: Add<Output = T> + Copy
+```
+
+Now consider:
+
+```
+fn add_self(x: impl Add<Output = impl Copy> + Copy)
+```
+
+You can see the duplication here and the overwhelming confusion that both the `impl Trait` will
```

Comment:
This argument is entirely subjective. I personally don't find this confusing. I'm taking "some type that can be added and copied. The result of the addition is some type that can be copied".

"overwhelming confusion" feels like unnecessary rhetoric to me.

Reply:
You’re right, it’s too much subjective. I’m removing it.